cmd/containerd/command/service_windows.go

```diff
@@ -0,0 +1,467 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package command
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"time"
+	"unsafe"
+
+	"github.com/containerd/containerd/services/server"
+	"github.com/sirupsen/logrus"
+	"github.com/urfave/cli"
+	"golang.org/x/sys/windows"
+	"golang.org/x/sys/windows/svc"
+	"golang.org/x/sys/windows/svc/debug"
+	"golang.org/x/sys/windows/svc/eventlog"
+	"golang.org/x/sys/windows/svc/mgr"
+)
+
+var (
+	serviceNameFlag       string
+	registerServiceFlag   bool
+	unregisterServiceFlag bool
+	runServiceFlag        bool
+
+	setStdHandle = windows.NewLazySystemDLL("kernel32.dll").NewProc("SetStdHandle")
+	oldStderr    windows.Handle
+	panicFile    *os.File
+
+	service *handler
+)
+
+const (
+	// These should match the values in event_messages.mc.
+	eventInfo  = 1
+	eventWarn  = 1
+	eventError = 1
+	eventDebug = 2
+	eventPanic = 3
+	eventFatal = 4
+
+	eventExtraOffset = 10 // Add this to any event to get a string that supports extended data
+)
+
+// serviceFlags returns an array of flags for configuring containerd to run
+// as a Windows service under control of SCM.
+func serviceFlags() []cli.Flag {
+	return []cli.Flag{
+		cli.StringFlag{
+			Name:  "service-name",
+			Usage: "Set the Windows service name",
+			Value: "containerd",
+		},
+		cli.BoolFlag{
+			Name:  "register-service",
+			Usage: "Register the service and exit",
+		},
+		cli.BoolFlag{
+			Name:  "unregister-service",
+			Usage: "Unregister the service and exit",
+		},
+		cli.BoolFlag{
+			Name:   "run-service",
+			Usage:  "",
+			Hidden: true,
+		},
+	}
+}
+
+// applyPlatformFlags applys platform-specific flags.
+func applyPlatformFlags(context *cli.Context) {
+
+	if s := context.GlobalString("service-name"); s != "" {
+		serviceNameFlag = s
+	}
+	for _, v := range []struct {
+		name string
+		d    *bool
+	}{
+		{
+			name: "register-service",
+			d:    &registerServiceFlag,
+		},
+		{
+			name: "unregister-service",
+			d:    &unregisterServiceFlag,
+		},
+		{
+			name: "run-service",
+			d:    &runServiceFlag,
+		},
+	} {
+		*v.d = context.GlobalBool(v.name)
+	}
+}
+
+type handler struct {
```

Comment:
I'm sure this logrus hook is the same as Docker right? Should we abstract this as a package and just let both share it? Seems like 100% common code.

Reply:
As discussed offline, it makes sense to abstract this once we have tracelog support in the future.

Reply:
Sounds good.